### PR TITLE
Make TabPy pip package

### PR DESCRIPTION
**Pull request by 0golovatyi**
With this change, TabPy is installed and executed as pip package.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -46,32 +51,29 @@ be able to work on TabPy changes:
     cd TabPy
     ```
 
-Before making any code changes run environment setup script.
-For Windows run this command from the repository root folder:
+4. Registeg TabPy repo as a pip package:
```
Comment (johng42) [inline CONTRIBUTING.md]: Nit: typo   -  Register TabPy repo as a pip package.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -46,32 +51,29 @@ be able to work on TabPy changes:
...
 
-```sh
-utils\set_env.cmd
-```
+    ```sh
+    pip install -e .
```
Comment (johng42) [inline CONTRIBUTING.md]: missing argument (pip install tabpy -e  ?)

File: docs/server-config.md [hunk 3]
```diff
@@ -4,47 +4,146 @@
...
+tabpy --config=path/to/my/config/file.conf
+```
 
 The default config file is provided to show you the default values but does not
 need to be present to run TabPy.
 
+### Configuration File Content
+
+Configuration file consists of settings for TabPy itself and Python logger
+settings. It is possible to only set parameters which value should be different
```
Comment (johng42) [inline docs/server-config.md]: the second sentence (It is possible to only...) doesn't make much sense.
Should it be
You should only need to set parameters if you need different values than the defaults.

?

File: docs/server-config.md [hunk 4]
```diff
@@ -4,47 +4,146 @@
...
+   value - `tabpy/tabpy_server` subfolder in TabPy package folder.
+- `TABPY_STATIC_PATH` - location of static files (index.html page) for
+  TabPy instance. Default value - `tabpy/tabpy_server/static` subfolder in
+  TabPy package folder.
+- `TABPY_PWD_FILE` - path to password file. Setting up this parameter
+  makes TabPy require credentials with HTTP(S) requests. More details about
+  authentication can be found in [Authentication](#authentication)
+  section. Default value - not set.
+- `TABPY_TRANSFER_PROTOCOL` - transfer protocol. Default value - `http`. If
+  set to `http` two additional parameters have to be specified:
```
Comment (johng42) [inline docs/server-config.md]: set to 'https' two additional ...

File: tabpy/tabpy_server/common/default.conf [hunk 5]
```diff
@@ -1,10 +1,10 @@
...
+# TABPY_QUERY_OBJECT_PATH = /tmp/query_objects
+# TABPY_PORT = 9004
+# TABPY_STATE_PATH = ./tabpy/tabpy_server
 
 # Where static pages live
-TABPY_STATIC_PATH = ./tabpy-server/tabpy_server/static
+# TABPY_STATIC_PATH = ./tabpy/tabpy_server/static
 
 # For how to configure TabPy authentication read
 # docs/server-config.md.
```
Comment (johng42) [inline tabpy/tabpy_server/common/default.conf]: this should be updated to 
[Authentication](#authentication) 
instead of server-config.md

File: CONTRIBUTING.md (hunk 1 above)
Comment (0golovatyi) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 2 above)
Comment (0golovatyi) [reply CONTRIBUTING.md]: no, this is actually correct - pip discovers package in the current folder as it has setup.py now.

File: docs/server-config.md (hunk 3 above)
Comment (0golovatyi) [reply docs/server-config.md]: fixed

File: docs/server-config.md (hunk 4 above)
Comment (0golovatyi) [reply docs/server-config.md]: fixed

File: tabpy/tabpy_server/common/default.conf (hunk 5 above)
Comment (0golovatyi) [reply tabpy/tabpy_server/common/default.conf]: This file is just plaintext and links won't work in it. Authentication explained in the server-config.md.

File: CONTRIBUTING.md (hunk 2 above)
Comment (johng42) [reply CONTRIBUTING.md]: TIL: Sweet!

File: .coveragerc [hunk 6]
```diff
@@ -4,4 +4,10 @@ exclude_lines =
     if __name__ == .__main__.:
 
 # Only show one number after decimal point in report.
-precision = 1
\ No newline at end of file
+precision = 1
+
+[run]
+omit =
+    models/
```
Comment (jnegara) [inline .coveragerc]: We have tests covering models, why are we omitting them from coverage reports?

File: .gitattributes [hunk 7]
```diff
@@ -1 +1 @@
-tabpy-server/tabpy_server/_version.py export-subst
+
```
Comment (jnegara) [inline .gitattributes]: Can we delete this file?

File: tabpy/models/scripts/SentimentAnalysis.py [hunk 8]
```diff
@@ -3,8 +3,16 @@
...
 from pathlib import Path
-sys.path.append(str(Path(__file__).resolve().parent.parent.parent / 'models'))
-from utils import setup_utils
+from tabpy.models.utils import setup_utils
+
+
+import ssl
+try:
+    _ctx = ssl._create_unverified_context
+except AttrubuteError:
```
Comment (jnegara) [inline tabpy/models/scripts/SentimentAnalysis.py]: `AttributeError`

Comment (jnegara) [reply tabpy/models/scripts/SentimentAnalysis.py]: I assume this code path hasn't been tested

File: tabpy/tabpy_server/app/app.py [hunk 9]
```diff
@@ -171,28 +173,37 @@ def set_parameter(settings_key,
                           config_key,
                           default_val=None,
                           check_env_var=False):
+            key_set = False
```
Comment (jnegara) [inline tabpy/tabpy_server/app/app.py]: why use a flag instead of else ifs? this variable is confusing, I expected it to be a set not a bool

File: tests/integration/test_deploy_model_ssl_on_auth_on.py [hunk 10]
```diff
@@ -19,10 +18,12 @@ def _get_pwd_file(self) -> str:
...
 
     def test_deploy_ssl_on_auth_on(self):
-        models = ['PCA', 'Sentiment%20Analysis', "ttest"]
-        path = str(Path('models', 'setup.py'))
-        p = subprocess.run([self.py, path, self._get_config_file_name()],
-                           input=b'user1\nP@ssw0rd\n')
+        # Uncomment the following line to preserve
+        # test case output and other files (config, state, ect.)
+        # in system temp folder.
+        # self.set_delete_temp_folder(False)
```
Comment (jnegara) [inline tests/integration/test_deploy_model_ssl_on_auth_on.py]: do we care about this being customizable? surely we're not going to advertise to users that they install Python via pip, navigate into site packages, and then uncomment a line of code in order to preserve temp files

File: .coveragerc (hunk 6 above)
Comment (0golovatyi) [reply .coveragerc]: they are not unit tests

File: .gitattributes (hunk 7 above)
Comment (0golovatyi) [reply .gitattributes]: sure

File: .vscode/settings.json [hunk 11]
```diff
@@ -13,10 +13,11 @@
...
     "python.testing.pytestArgs": [
-        "."
+        "tests"
     ],
     "python.testing.unittestEnabled": false,
     "python.testing.nosetestsEnabled": false,
     "python.testing.pytestEnabled": true,
-    "python.linting.pep8Enabled": true
+    "python.linting.pep8Enabled": true,
+    "python.pythonPath": "C:\\Users\\ogolovatyi\\AppData\\Local\\conda\\conda\\envs\\python 3.6\\python.exe"
```
Comment (0golovatyi) [inline .vscode/settings.json]: CLean this up!

File: docs/server-config.md [hunk 12]
```diff
@@ -4,47 +4,146 @@
...
+- `TABPY_EVALUATE_TIMEOUT` - script evaluation timeout in seconds. Default
+  value - `30`.
+
+### Configuration File Example
+
+```toml
+[TabPy]
+# TABPY_QUERY_OBJECT_PATH = /tmp/query_objects
+# TABPY_PORT = 9004
+# TABPY_STATE_PATH = ./tabpy/tabpy_server
```
Comment (jnegara) [inline docs/server-config.md]: Let's document that these should be absolute paths not relative paths.

File: setup.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+            'python-dateutil',
+            'requests',
+            'singledispatch',
+            'six',
+            'tornado',
+            'urllib3<1.25,>=1.21.1'
+        ],
+        entry_points={
+            'console_scripts': [
+                'tabpy=tabpy.tabpy:main',
```
Comment (jnegara) [inline setup.py]: Can we add the script for adding users and hashed passwords to a file?

File: tabpy/models/scripts/SentimentAnalysis.py (hunk 8 above)
Comment (0golovatyi) [reply tabpy/models/scripts/SentimentAnalysis.py]: No, it hasn't. Removing try/catch - it better fail with exception IMO.

File: tabpy/tabpy_server/app/app.py (hunk 9 above)
Comment (0golovatyi) [reply tabpy/tabpy_server/app/app.py]: nested if's - this is why. renaming to key_is_set

File: tests/integration/test_deploy_model_ssl_on_auth_on.py (hunk 10 above)
Comment (0golovatyi) [reply tests/integration/test_deploy_model_ssl_on_auth_on.py]: Correct

File: docs/server-config.md (hunk 12 above)
Comment (0golovatyi) [reply docs/server-config.md]: fixed

File: setup.py (hunk 13 above)
Comment (0golovatyi) [reply setup.py]: added with the new set of changes

File: docs/server-config.md [hunk 14]
```diff
@@ -4,47 +4,149 @@
...
 <!-- markdownlint-enable MD004 -->
 
-Default settings for TabPy may be viewed in the
-tabpy_server/common/default.conf file. This file also contains
-commented examples of how to set up your TabPy server to only
-serve HTTPS traffic and enable authentication.
+## Custom Settings
 
-Change settings by:
+TabPy starts with set of default settings unless settings are provided via
```
Comment (sbabayan) [inline docs/server-config.md]: TabPy starts with a set of default settings unless settings are provided via environment variables or with a config file.

File: docs/server-config.md [hunk 15]
```diff
@@ -4,47 +4,149 @@
...
-3. Specifying your own config file as a command line parameter.
-   - i.e. Running the command:
+Configuration parameters can be updated with:
 
-     ```sh
-     python tabpy.py --config=path\to\my\config
-     ```
+1. Adding environment variables - set the environment variable as required by
+   your Operating System. When creating environment variables, use the same
+   name as is in the config file as an environment variable.
```
Comment (sbabayan) [inline docs/server-config.md]: When creating environment variables, use the same name for your environment variable as specified in the config file.

File: docs/server-config.md [hunk 16]
```diff
@@ -4,47 +4,149 @@
...
+2. Specifying a parameter in a config file (enviroment variable value overwrites
```
Comment (sbabayan) [inline docs/server-config.md]: environment*

File: docs/server-config.md [hunk 17]
```diff
@@ -4,47 +4,149 @@
...
 The default config file is provided to show you the default values but does not
 need to be present to run TabPy.
 
+### Configuration File Content
+
+Configuration file consists of settings for TabPy itself and Python logger
+settings. You should only set parameters if you need different values than
+the defaults.
+
+TabPy parameters explained below and for the logger documentation
```
Comment (sbabayan) [inline docs/server-config.md]: TabPy parameters explained below, the logger documentation can be found at

File: docs/server-config.md [hunk 18]
```diff
@@ -4,47 +4,149 @@
...
+- `TABPY_PWD_FILE` - path to password file. Setting up this parameter
+  makes TabPy require credentials with HTTP(S) requests. More details about
+  authentication can be found in [Authentication](#authentication)
+  section. Default value - not set.
+- `TABPY_TRANSFER_PROTOCOL` - transfer protocol. Default value - `http`. If
+  set to `https` two additional parameters have to be specified:
+  `TABPY_CERTIFICATE_FILE` and `TABPY_KEY_FILE`. More details for how to
+  configure TabPy for HTTPS are at [Configuring HTTP vs HTTPS]
+  (#configuring-http-vs-https) section.
+- `TABPY_CERTIFICATE_FILE` to certificate file to run TabPy with. Only used
```
Comment (sbabayan) [inline docs/server-config.md]: the certificate file to run TabPy with

File: docs/server-config.md [hunk 19]
```diff
@@ -4,47 +4,149 @@
...
+  with `TABPY_TRANSFER_PROTOCOL` set to `https`. Default value - not set.
```
Comment (sbabayan) [inline docs/server-config.md]: when `TABPY_TRANSFER_PROTOCOL` is set to

File: docs/server-config.md [hunk 20]
```diff
@@ -4,47 +4,149 @@
...
+- `TABPY_KEY_FILE` to private key file to run TabPy with. Only used
```
Comment (sbabayan) [inline docs/server-config.md]: the private key file

File: docs/server-config.md [hunk 21]
```diff
@@ -4,47 +4,149 @@
...
+  with `TABPY_TRANSFER_PROTOCOL` set to `https`. Default value - not set.
```
Comment (sbabayan) [inline docs/server-config.md]: when `TABPY_TRANSFER_PROTOCOL` is set to

File: docs/server-config.md [hunk 22]
```diff
@@ -4,47 +4,149 @@
...
+# TABPY_KEY_FILE = path/to/key/file.key
+
+# Log additional request details including caller IP, full URL, client
+# end user info if provided.
+# TABPY_LOG_DETAILS = true
+
+# Configure how long a custom script provided to the /evaluate method
+# will run before throwing a TimeoutError.
+# The value should be a float representing the timeout time in seconds.
+#TABPY_EVALUATE_TIMEOUT = 30
```
Comment (sbabayan) [inline docs/server-config.md]: Please put a space between # and TABPY_EVALUATE_TIMEOUT it is really bothering me that all the other lines have spaces but this one doesn't

File: docs/server-install.md [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+## TabPy Installation
+
+To install TabPy on to an environment `pip` needs to be installed and
+updated first:
+
+```sh
+python -m pip install --upgrade pip
+```
+
+Now TabPy can be install as a package:
```
Comment (sbabayan) [inline docs/server-install.md]: installed*

File: docs/server-install.md [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+
+```sh
+pip install tabpy
+```
+
+## Starting TabPy
+
+To start TabPy with default setting run the following command:
```
Comment (sbabayan) [inline docs/server-install.md]: with default settings*

File: docs/server-install.md [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+
+```sh
+tabpy
+```
+
+To run TabPy with custom setting create config file with parameters
```
Comment (sbabayan) [inline docs/server-install.md]: settings*

File: docs/tabpy-tools.md [hunk 26]
```diff
@@ -161,34 +161,41 @@ client.remove('WillItDefault')
...
 
 ```sh
 "Successfully deployed PCA"
 ```
 
-If you would like to deploy additional models using the deploy script, you can
-copy any python file to the `./models/scripts` directory and modify setup.py to
-include all necessary packages when installing dependencies, or alternatively install
+Use code in [`tabpy/models/scripts`](../tabpy/models/scripts)
+as an example how to create a model and
```
Comment (sbabayan) [inline docs/tabpy-tools.md]: as an example of how to create a model*

File: tabpy/models/scripts/SentimentAnalysis.py [hunk 27]
```diff
@@ -3,8 +3,12 @@
 from nltk.sentiment.vader import SentimentIntensityAnalyzer
 import sys
 from pathlib import Path
-sys.path.append(str(Path(__file__).resolve().parent.parent.parent / 'models'))
-from utils import setup_utils
+from tabpy.models.utils import setup_utils
+
+
+import ssl
```
Comment (sbabayan) [inline tabpy/models/scripts/SentimentAnalysis.py]: Why is this necessary only for Sentiment Analysis?

File: tabpy/models/utils/setup_utils.py [hunk 28]
```diff
@@ -34,7 +41,7 @@ def get_creds():
     return [user, passwd]
 
 
-def main(funcName, func, funcDescription):
+def deploy_model(funcName, func, funcDescription):
     # running from setup.py
```
Comment (sbabayan) [inline tabpy/models/utils/setup_utils.py]: update comment to say deploy_models.py

File: tabpy/tabpy_server/common/default.conf [hunk 29]
```diff
@@ -1,13 +1,13 @@
 [TabPy]
-TABPY_QUERY_OBJECT_PATH = /tmp/query_objects
-TABPY_PORT = 9004
-TABPY_STATE_PATH = ./tabpy-server/tabpy_server
+# TABPY_QUERY_OBJECT_PATH = /tmp/query_objects
```
Comment (sbabayan) [inline tabpy/tabpy_server/common/default.conf]: why even specify a default.conf if it's all commented out and we already have documentation on how to create your own custom.conf?

File: docs/server-config.md (hunk 14 above)
Comment (0golovatyi) [reply docs/server-config.md]: fixed

File: docs/server-config.md (hunk 15 above)
Comment (0golovatyi) [reply docs/server-config.md]: fixed

File: docs/server-config.md (hunk 18 above)
Comment (0golovatyi) [reply docs/server-config.md]: fixed

File: docs/server-config.md (hunk 17 above)
Comment (0golovatyi) [reply docs/server-config.md]: fixed

File: docs/server-install.md (hunk 25 above)
Comment (0golovatyi) [reply docs/server-install.md]: fixed

File: tabpy/models/scripts/SentimentAnalysis.py (hunk 27 above)
Comment (0golovatyi) [reply tabpy/models/scripts/SentimentAnalysis.py]: nltk package requires dictionaries which are not installed with the package. And the download is done with HTTPS where the server has untrusted cert. To prevent connection failure the code tells to not validate a cert.

File: tabpy/models/utils/setup_utils.py (hunk 28 above)
Comment (0golovatyi) [reply tabpy/models/utils/setup_utils.py]: fixed

File: tabpy/tabpy_server/common/default.conf (hunk 29 above)
Comment (0golovatyi) [reply tabpy/tabpy_server/common/default.conf]: I simply didn't want to add change to completely remove the default config file. If we do that we will need some template for the config anyway. 

Probably should do with some next improvement.

File: tabpy/models/deploy_models.py [hunk 30]
```diff
@@ -48,3 +48,6 @@ def install_dependencies(packages):
     for filename in os.listdir(directory):
         subprocess.run([py, f'{directory}/{filename}', config_file_path]
```
Comment (sbabayan) [inline tabpy/models/deploy_models.py]: Did you smoke test this with a custom config? I'm not sure about specifying absolute vs relative path for the custom config anymore

**Comment by jnegara** [comment]
Is this still WIP?

File: tabpy/models/deploy_models.py (hunk 30 above)
Comment (0golovatyi) [reply tabpy/models/deploy_models.py]: I updated deploy_and_evaluate integration test to use the custom config.

Comment (sbabayan) [reply tabpy/models/deploy_models.py]: I see the tests now modify the state.ini file, shouldn't they write to a different state file so that running the tests doesn't add the models endpoints?

File: tabpy/tabpy_server/state.ini [hunk 31]
```diff
@@ -7,9 +7,15 @@ Access-Control-Allow-Headers =
 Access-Control-Allow-Methods = 
 
 [Query Objects Service Versions]
+PCA = {"description": "Returns the specified principal component", "type": "model", "version": 2, "dependencies": [], "target": null, "creation_time": 1565729597, "last_modified_time": 1565730202, "schema": null}
```
Comment (sbabayan) [inline tabpy/tabpy_server/state.ini]: remove this from state.ini pls

Comment (0golovatyi) [reply tabpy/tabpy_server/state.ini]: This file shouldn't be in the repo at all, deleting.

File: tabpy/models/deploy_models.py (hunk 30 above)
Comment (0golovatyi) [reply tabpy/models/deploy_models.py]: state.ini is not in the repo anymore and is created as a new file for each test from state.ini.template.

File: docs/tabpy-tools.md (hunk 26 above)
Comment (0golovatyi) [reply docs/tabpy-tools.md]: fixed
